### PR TITLE
Confidential transaction #130

### PR DESCRIPTION
**Pull request by hihiben**
##### Why:
Implement encryption and decryption for confidential transaction.
   
##### How:
The transaction to be encrypted is encrypted through AES with a pair of random generated key and iv. The key and iv will then encrypted by the given pubkey through ECIES(secp256k1 curve).  
##### Usage:
-  Add several member in `CTransaction` and `CMutableTransaction`
    - `std::string chex`
Keep the encrypted hex string of the confidential area  
    - `std::vector<CPubKey> pubKeys`
Keep the public key of encrypting the hex
    - `std::vector<std::string>` encryptedKeys
Keep the encrypted AES key and iv pairs 
-  rpc `sendfrom`
A optional json array of pubkey attached at the end of arguments. Only the nodes that own the keys is able to read vin, vout, nLockTime, type. 

##### TODO
-  Unit test
-  Lighter crypt library for ECIES (currently entire lib cryptopp)

### CODE REVIEW

File: src/primitives/transaction.cpp [hunk 1]
```diff
@@ -174,3 +242,73 @@ std::string GetTypeName(tx_type type)
     }
 }
 
+bool CMutableTransaction::Encrypt(const std::vector<CPubKey>& vchPubKeys)
+{
+    if (vchPubKeys.empty())
+        return false;
+    if (!chex.empty())
+        return true;
```
Comment (lbyyy) [inline src/primitives/transaction.cpp:250]: why true?

File: src/main.cpp [hunk 2]
```diff
@@ -4012,7 +4040,7 @@ bool CheckBlock(const CBlock& block, CValidationState& state, bool fCheckPOW, bo
     // Check transactions
     BOOST_FOREACH(const CTransaction& tx, block.vtx)
     {
-        if (!ExistInPool(tx) && !CheckTransaction(tx, state))
+        if (!ExistInPool(tx) && !tx.IsEncrypted() && !CheckTransaction(tx, state))
```
Comment (lbyyy) [inline src/main.cpp:4043]: 這樣是check block時只要看不懂就return error?

File: src/primitives/transaction.cpp (hunk 1 above)
Comment (hihiben) [reply src/primitives/transaction.cpp:250]: 如果已經有chex的部分的話代表這筆交易已經被encrypt過了

File: src/main.cpp (hunk 2 above)
Comment (hihiben) [reply src/main.cpp:4043]: 是他如果被加密過的話直接跳過了不跑後面的CheckTransaction

**Review by lbyyy — APPROVED** [review]

**Pull request merged**